### DISCUSSION TITLE
In CI/CD, what goes into $tpp_token?

### DISCUSSION THREAD

**Discussion by bannbay** [Venafi/venafips#223] [Q&A] [answered]
How do I define tpp_token to send as a parameter as session for a command like Find-VenafiCertificate?

Find-VenafiCertificate -VenafiSession $VenafiSession doesn't work. 
This is how I generate the token/session - 
New-VenafiSession -Credential $credential -Server Venafi.org.com -ClientId  vcert-cli -scope @{'certificate'='manage,revoke';'configuration'='manage';'ssh'='manage'}

I keep getting this error -

Please run New-VenafiSession or provide a VaaS key or TPP token.
At C:\Program Files\WindowsPowerShell\Modules\VenafiPS\5.8.1\Private\Test-VenafiSession.ps1:80 char:17
+ ...             throw 'Please run New-VenafiSession or provide a VaaS key ...

**Comment by gdbarron** [comment, marked as answer]
If New-VenafiSession is used as is, without -PassThru, there is no need to provide -VenafiSession to any function as the venafi session is saved to a local session variable.

``` powershell
New-VenafiSession -Credential $credential -Server Venafi.org.com -ClientId vcert-cli -scope @{'certificate'='manage,revoke';'configuration'='manage';'ssh'='manage'}
Find-VenafiCertificate
```

Optionally, you can add -PassThru to New-VenafiSession, set the output to a variable and then pass that to Find-VenafiCertificate as -VenafiSession.  In this case, I recommend using a variable name other than $VenafiSession.

``` powershell
$sess = New-VenafiSession -Credential $credential -Server Venafi.org.com -ClientId vcert-cli -scope @{'certificate'='manage,revoke';'configuration'='manage';'ssh'='manage'} -PassThru
Find-VenafiCertificate -VenafiSession $sess
```

**Reply by bannbay** [reply]
Without -Passthru, I get 
Please run New-VenafiSession or provide a VaaS key or TPP token.
At C:\Program Files\WindowsPowerShell\Modules\VenafiPS\5.8.1\Private\Test-VenafiSession.ps1:80 char:17

With -Passthru I get

Token                : @{Server=https://venafi.org.com; 
                       AccessToken=System.Management.Automation.PSCredential; 
                       RefreshToken=System.Management.Automation.PSCredential; Scope=System.Collections.Hashtable; 
                       Identity=AD+org.com:753a01e74dc8d0469e8d82a1bf5cb4f1; TokenType=Bearer; 
                       ClientId=vcert-cli; Expires=12/20/2023 2:21:51 PM; RefreshExpires=9/20/2024 2:21:51 PM}

E:\VSTS\agent\_work\r10\a\ExpiringCerts.ps1 : Cannot validate argument on 
parameter 'VenafiSession'. The argument is null or empty. Provide an argument that is not null or empty, and then try
the command again.


This is my script -

$VSession  = New-VenafiSession -serverurl venafi.org.com -Credential  $pscredential -ClientId 'vcert-cli' -scope @{'certificate'='manage,revoke';'configuration'='manage';'ssh'='manage'}  -PassThru
$VSession
Find-VenafiCertificate -venafisession $VSession | select Name, ParentPath | export-csv $workingpath\export.csv

**Reply by gdbarron** [reply]
If you run New-VenafiSession without -PassThru, $VenafiSession should be stored in the session.  Do you see something similar to this when you run `$VenafiSession`?
![image](https://github.com/Venafi/VenafiPS/assets/11862024/a32f7eb9-1cca-4a72-a901-b1ea41b9cb77)
